### PR TITLE
Split engine into two assemblies

### LINKED ISSUE #684: Split engine into upper and lower parts
I did PR #580 "Split engine spike" as a spike, i.e. code to prove a concept, intended to be discarded and re-developed more carefully under test. Based on the spike, I'm writing this issue as a proposal to go ahead and move to the actual implementation. Comments on the spike already indicated a general consensus that this would be a desirable evolution of the engine. Before starting to code for real I'd like to have some review of what I propose here.

PROPOSAL

Split the current engine assembly nto two parts, a lower and an upper part, with the upper part making use of the lower part. The upper and lower parts would be built for potentially different targets - more for the lower part than the upper.

I will continue to call the upper part the `engine`. It needs to be built for platforms on which developers actually edit, compile and test their programs. Initially, I'd retain .NET 2.0 and .NET Core 2.0 for this purpose.

I'm calling the lower part the "agent core" for the moment. It needs to be built for any targets on which we want the tests to run, i.e. all current targets. If, however, we are about to remove any existing targets, it would be convenient to do it before starting this work.

APPROACH

My preference is to do this as a series of PRs, including tests, so that the change takes place in smaller, easier to review steps. If preferred, I can use a "big bang" approach, however.

REVIEW

I'm looking for a review of the general direction before starting work. Obviously, the details will be looked at again in code review but I'd like to know that this is wanted before spending time on it.

### PR DESCRIPTION
**Pull request by CharliePoole**
Fixes #684 

This is an alternative approach to issue #684, as compared to that of PR #689.

I've brought this up to the same stage of (partial) completion as #689 so they can be compared. There's lots more to do and I'd like to pick one approach before continuing.

As explained elsewhere, I think this PR, which creates only a single new assembly, is to be preferred at the moment. Starting from scratch, I might do it otherwise, but we're pretty far from that. Creating a single new assembly leaves the door open for further splitting if it should be called for.

Both this PR and #689 only deal with splitting the engine, which is still referenced as if it were a single entity. Once the split is complete, then more work is needed to actually use the lower part by itself in the agent assemblies.

I had hoped this could be done as a set of small PRs. But that would require reviewing one bit of it without knowing what comes next. I'm happy to merge either this or #689 as a whole or up to a certain point. Both PRs are probably best reviewed one commit at a time.

FYI @nunit/engine-team - I'll be spending tomorrow flying back to the US and probably won't get back to this until the day after. Let me know which PR you think I should continue.

### CODE REVIEW

**Comment by ChrisMaddock** [comment]
Hi @CharliePoole,

I've had a quick look through both - long story short, I agree. The reasoning for the division in the first PR was unclear to me even, as someone who's familiar with the codebase. I'm sure the reasoning behind this would make sense if I started trying to move classes myself, and found out more about the interdependencies - but the version in this PR makes more sense to me to first come to.

**Comment by ChrisMaddock** [comment]
Hi Charlie - I've had a quick look through both, and long-story-short, I agree. The division in the first version is unclear to me, even as someone who's familiar with the codebase. I'm sure that would start to make sense if I started to pull things apart and learnt more about the inter-dependencies, but second version feels much more approachable to me.

> I had hoped this could be done as a set of small PRs. But that would require reviewing one bit of it without knowing what comes next

I prefer working like that myself, if it's possible to cover a quick 'what and why' in the PR description. Whatever works best, though.

**Comment by CharliePoole** [comment]
@ChrisMaddock I'll continue in this way - splitting into two assemblies - and I'll eventually delete the old ones (both #689 and #580).

Regarding small PRs, let's use this one as a roadmap. I've added @jnm2 to the requested reviewers. @jnm2 can you figure out where in the series of commits this starts interfering with what you are doing? It's trivial to break this big PR up if we want.

Meanwhile, I'm continuing in this path, using some of the jet-lag energy I have after flying home yesterday.

**Comment by CharliePoole** [comment]
This is ready for review now.

All the commits up to the last one simply move code to the new assembly and make adjustments so that the tests keep passing. The final commit is the real meat of it, since it causes the agents to reference the `nunit.engine.core` assembly and use `CoreEngine`, rather than `nunit.engine` and `TestEngine`.

There are a few more things that could be done, but this PR is as big as I want to go. I'm also willing to split it up into several PRs if anyone sees value in that.

**Comment by CharliePoole** [comment]
@nunit/engine-team I'd appreciate somebody - or even everybody - chiming in on this. I'm blocked on about four other things while waiting for clarification on which way you want to go. To be clear, even if it isn't quite ready to be merged, I need to know what direction folks want to go before I'll be able to work on those other things. So if you don't have time for a detailed review, that's OK, but your views on (1) whether to break up the engine this way, (2) whether to have two pieces or three and (3) whether I need to break up this PR to accommodate other work in progress would really help.

**Comment by ChrisMaddock** [comment]
Hi Charlie. Have left my thoughts on direction above. This PR looks fine to review as is to me, it's a lot of changes, but the vast majority are just file moves

File: src/NUnitEngine/nunit.engine.tests/Integration/RunnerInDirectoryWithoutFramework.cs [hunk 1]
```diff
@@ -16,7 +16,7 @@ internal sealed class RunnerInDirectoryWithoutFramework : IDisposable
 
         public RunnerInDirectoryWithoutFramework()
         {
-            directory = new DirectoryWithNeededAssemblies("nunit3-console", "nunit-agent", "nunit-agent-x86");
+            directory = new DirectoryWithNeededAssemblies("nunit3-console", "nunit.engine", "nunit-agent", "nunit-agent-x86");
```
Comment (ChrisMaddock) [inline src/NUnitEngine/nunit.engine.tests/Integration/RunnerInDirectoryWithoutFramework.cs]: Heh - took me a while to figure out why this change was needed. Guess that's confirmation that the structure is right! 😄

File: src/NUnitEngine/nunit.engine/nunit.engine.csproj [hunk 2]
```diff
@@ -21,9 +21,11 @@
   <ItemGroup>
     <Compile Include="..\EngineVersion.cs" LinkBase="Properties" />
     <Compile Include="..\nunit-agent\AgentExitCodes.cs" LinkBase="Agents" />
+    <Compile Include="..\nunit.engine.core\Internal\ExceptionHelper.cs" Link="Internal\ExceptionHelper.cs" />
```
Comment (ChrisMaddock) [inline src/NUnitEngine/nunit.engine/nunit.engine.csproj]: As above - can this just reference this class directly, rather than the link, to be consistent with the other Util classes?

File: src/NUnitEngine/nunit.engine.core/CoreEngine.cs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        /// Initialize the engine. This includes initializing mono addins,
+        /// setting the trace level and creating the standard set of services
+        /// used in the Engine.
+        ///
+        /// This interface is not normally called by user code. Programs linking
+        /// only to the nunit.engine.api assembly are given a
+        /// pre-initialized instance of TestEngine. Programs
+        /// that link directly to nunit.engine usually do so
+        /// in order to perform custom initialization.
+        /// </summary>
```
[the remark this answers is not in the document]
Comment (ChrisMaddock) [reply src/NUnitEngine/nunit.engine.core/CoreEngine.cs]: Waaaaait, the agent's still _do_ initialize services themselves. So from what I see, no production code yet adds services here. 

Should that bit come out? The duplication did confuse me...

File: src/NUnitEngine/nunit.engine.core/CoreEngine.cs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        //            Initialize();
+
+        //        return Services;
+        //    }
+        //}
+
+        /// <summary>
+        /// Initialize the engine. This includes initializing mono addins,
+        /// setting the trace level and creating the standard set of services
+        /// used in the Engine.
```
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.core/CoreEngine.cs]: Very old! Fixed.

**Comment by CharliePoole** [comment]
@ChrisMaddock Thanks for your comments. I'll go through and make fixes this evening - moving house today! 

Regarding putting some of the things with the agent itself... I envision a future where we have multiple agents, which is why I was playing around with the idea of a third library to hold stuff that is used in common by agents. I saw this change as the minimum needed to be done all at once to have something that made sense but I didn't want to add more.

I'll definitely be going further with this approach - perhaps in my own build, or simultaneously with NUnit if we end up going the same way. For what it's worth, my notion of an "agent" is that it encapsulates the "where we run" part of things along with the particular type of communication used to send commands and receive events and results. Right now, we only have one true agent, which runs in a separate process on the same machine and uses remoting to communicate. IME the big benefit of splitting out agents and making them as simple as possible is that we don't have to limit how many agents we support.

**Comment by ChrisMaddock** [comment]
'CharliwPoole Oh

**Comment by ChrisMaddock** [comment]
@CharliePoole Oh! We'll need to update all the packaging for this too, right? NuGet, Choco and the msi will all definitely be affected - not sure about how the zip is gathered in Cake, off the top of my head.

**Comment by CharliePoole** [comment]
Agreed. Are you OK with that as a separate PR?

**Comment by ChrisMaddock** [comment]
I'd _prefer_ to keep master releasable with this one (i.e. all in one PR) - but happy to go with two separate PR's, if you feel this is getting laborious to work with. 🙂

**Comment by CharliePoole** [comment]
Amount of work for me is about the same but one PR gives re-reviews of the whole thing. Also, current PR will likely cause merge conflicts for others whereas packaging probably won't.

**Comment by CharliePoole** [comment]
Easiest packaging will be to retain same packages for this release. Consider adding separate core package later.

**Comment by ChrisMaddock** [comment]
Absolutely agree on not adding 'core package' - was just thinking the assembly needs to be included in the current packages, otherwise they won't work! I think it would be better to do that in the same PR if that's ok with you - it's only a few lines, and easy to review separately. 🙂 

At this point, I'm planning to rebase other PR's over this one, rather than the other way round - as I think that'll be easier to do.

In terms of reviews on this, I'm happy to merge this just with mine and yours reviews, if you would be - I appreciate others are pretty busy at the moment, and this is a lot to look through! I'm happy that enough seen/commented on #684 and #580 to agree with the approach - and the details are pretty clear-cut here if we're aiming for 'minimum change', and can always be changed in future. 👍

**Comment by CharliePoole** [comment]
OK... let's do that. Ill-treatment to finish today, Pacific time.

**Comment by ChrisMaddock** [comment]
No rush on my part! 🙂

File: src/NUnitEngine/nunit-agent/Program.cs [hunk 5]
```diff
@@ -111,11 +111,12 @@ public static void Main(string[] args)
               log.Warning("Failed to restore the COMPLUS_Version variable. " + ex.Message); // Proceed with running tests anyway
             }
 
-            // Create TestEngine - this program is
+            // Create CoreEngine - this program is
             // conceptually part of  the engine and
             // can access its internals as needed.
```
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitEngine/nunit-agent/Program.cs]: Note that the core engine is using four services, including extension service. I think you'll see they are all needed. If services become extensions at that point, then that's how they would have to be loaded.

File: src/NUnitEngine/nunit-agent/nunit-agent-x86.csproj [hunk 6]
```diff
@@ -14,7 +14,7 @@
   </ItemGroup>
   <ItemGroup>
     <Compile Include="..\EngineVersion.cs" LinkBase="Properties" />
-    <Compile Include="..\nunit.engine\Internal\ExceptionHelper.cs" Link="ExceptionHelper.cs" />
+    <Compile Include="..\nunit.engine.core\Internal\ExceptionHelper.cs" Link="ExceptionHelper.cs" />
```
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitEngine/nunit-agent/nunit-agent-x86.csproj]: It's tricky. The console runner would then have to directly reference the core engine directly. Otherwise we end up with duplication of the class in console tests. I followed that path for a bit, until it became too hairy and I reverted to shared source code for this particular helper.

File: src/NUnitEngine/nunit.engine.core/Services/ExtensionService.cs [hunk 7]
```diff
@@ -41,14 +41,16 @@ namespace NUnit.Engine.Services
...
-        static readonly Version ENGINE_VERSION = typeof(TestEngine).Assembly.GetName().Version;
+        static readonly Version ENGINE_VERSION = typeof(ExtensionService).Assembly.GetName().Version;
 
         private readonly List<ExtensionPoint> _extensionPoints = new List<ExtensionPoint>();
         private readonly Dictionary<string, ExtensionPoint> _pathIndex = new Dictionary<string, ExtensionPoint>();
 
         private readonly List<ExtensionNode> _extensions = new List<ExtensionNode>();
         private readonly List<ExtensionAssembly> _assemblies = new List<ExtensionAssembly>();
 
+        public IList<Assembly> RootAssemblies { get; } = new List<Assembly>();
```
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.core/Services/ExtensionService.cs]: Odd... it was used. Looking just at the code, it appears that extension points in the `nunit.engine.core` and `nunit.engine.api` will be found but not those in `nunit.engine`. But we know they are found because the tests pass!

Obviously, in the course of implementing this three times, I must have changed the approach. I'll verify and possibly remove this.

File: src/NUnitEngine/nunit.engine.core/nunit.engine.core.csproj [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    <Compile Include="Internal\ProcessUtils.cs" />
+    <Compile Include="Internal\ProvidedPathsAssemblyResolver.cs" />
+    <Compile Include="Internal\ResultHelper.cs" />
+    <Compile Include="Internal\RuntimeFrameworks\DotNetFrameworkLocator.cs" />
+    <Compile Include="Internal\ServerBase.cs" />
+    <Compile Include="Internal\TargetFrameworkHelper.cs" />
+    <Compile Include="Internal\TcpChannelUtils.cs" />
+    <Compile Include="Internal\TcpChannelUtils.ObservableServerChannelSinkProvider.cs" />
+    <Compile Include="Internal\TestPackageExtensions.cs" />
+  </ItemGroup>
```
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.core/nunit.engine.core.csproj]: The files get added when using VS 2017 to cut and paste them from one project to another. I figured on going back to this but didn't. I think we're done moving files now, so I could try removing them now.

File: src/NUnitEngine/nunit.engine.tests/Integration/RunnerInDirectoryWithoutFramework.cs (hunk 1 above)
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.tests/Integration/RunnerInDirectoryWithoutFramework.cs]: Yeah, @jnm2's code for copying the directories only goes one level deep in following references, because that's all that was needed at the time. Rather than change it, I opted to just add the engine explicitly, thereby picking up core.

File: src/NUnitEngine/nunit.engine/nunit.engine.csproj (hunk 2 above)
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine/nunit.engine.csproj]: See earlier comment.

File: src/NUnitEngine/nunit.engine.core/Services/ExtensionService.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.core/Services/ExtensionService.cs]: OK... I realized at some point that all our extension points are __currently__ defined in `nunit.engine.api` and `nunit.engine.core` (only one in the latter). I think `RootAssemblies` was needed in a three-way split, but not in a two-way split.

I think some way to specify roots to be searched for extension points will be needed at some point - for example if we want to allow extension points in nunit3-console. However, it could be removed now or just commented as being for future use. I'm OK either way.

File: src/NUnitEngine/nunit.engine.core/nunit.engine.core.csproj (hunk 8 above)
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitEngine/nunit.engine.core/nunit.engine.core.csproj]: Done.

**Comment by CharliePoole** [comment]
@ChrisMaddock Ready for another review. I added `nunit.engine.core` to all the packages with `nunit.engine` and verified the contents for everything but the msi, which I'm not set up to build locally. Someone should check that.

**Comment by jnm2** [comment]
> I'm particularly concerned about not messing up the long-running PR you are working on.

Thanks. I'm planning to redo it all though, just referencing my current PR as I go and trying to do it in as small of PRs as I can figure out. I hope you are able to feel free not to worry about the long-running PR. (Maybe I should close it for now?)
